### PR TITLE
Creates a LoggingChannelFactory.

### PR DESCRIPTION
**Pull request by jdcove2**
This PR creates a SeekableByteChannelFactory that wraps another SeekableByteChannelFactory and logs every method call on any SeekableByteChannels that are created. This is intended only for Development/Debugging purposes.

### CODE REVIEW

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class creates a SeekableByteChannelFactory for logging accesses to every method on the SeekableByteChannel. It
+ * is intended only for development/debugging purposes.
+ */
+public final class LoggingChannelFactory {
```
Comment (drivenflywheel) [inline src/main/java/emissary/core/channels/LoggingChannelFactory.java:16]: If this class is meant to be used only during debugging/development, maybe we should consider decorating it with a `@Deprecated` annotation that will generate compiler warnings if any usage gets committed?

**Review by drivenflywheel — APPROVED** [review]
Might want to consider adding a `@Deprecated` annotation to flag committed usages, but overall this looks good

**Review by smcgrattan — APPROVED** [review]

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                    bytesWritten);
+
+            return bytesWritten;
+        }
+
+        @Override
+        public long position() throws IOException {
+            final long position = seekableByteChannel.position();
+
+            log(logger, logStackTrace, "{} : position : r={}", prefix, position);
```
Comment (fbruton) [inline src/main/java/emissary/core/channels/LoggingChannelFactory.java:163]:
```suggestion
            log(logger, logStackTrace, "{} : position : p={}", prefix, position);
```

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            log(logger, logStackTrace, "{} : position : nP={}", prefix, newPosition);
+
+            return seekableByteChannel.position(newPosition);
+        }
+
+        @Override
+        public long size() throws IOException {
+            final long size = seekableByteChannel.size();
+
+            log(logger, logStackTrace, "{} : size : r={}", prefix, size);
```
Comment (fbruton) [inline src/main/java/emissary/core/channels/LoggingChannelFactory.java:179]:
```suggestion
            log(logger, logStackTrace, "{} : size : s={}", prefix, size);
```

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+        @Override
+        public int write(final ByteBuffer byteBuffer) throws IOException {
+            final int bbP = byteBuffer.position();
+            final int bbC = byteBuffer.capacity();
+            final long sbcP = seekableByteChannel.position();
+
+            final int bytesWritten = seekableByteChannel.write(byteBuffer);
+
+            log(logger, logStackTrace, "{} : write : bbP={} bbC={} sbcP={} : r={}", prefix, bbP, bbC, sbcP,
```
Comment (fbruton) [inline src/main/java/emissary/core/channels/LoggingChannelFactory.java:153]:
```suggestion
            log(logger, logStackTrace, "{} : write : bbP={} bbC={} sbcP={} : bW={}", prefix, bbP, bbC, sbcP,
```

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            this.logStackTrace = logStackTrace;
+
+            log(logger, logStackTrace, "{} : created : lST={}", prefix, logStackTrace);
+        }
+
+        @Override
+        public boolean isOpen() {
+            final boolean open = seekableByteChannel.isOpen();
+
+            log(logger, logStackTrace, "{} : isOpen : r={}", prefix, open);
```
Comment (fbruton) [inline src/main/java/emissary/core/channels/LoggingChannelFactory.java:120]:
```suggestion
            log(logger, logStackTrace, "{} : isOpen : open={}", prefix, open);
```

**Review by fbruton — CHANGES_REQUESTED** [review]

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java (hunk 2 above)
Comment (jdcove2) [reply src/main/java/emissary/core/channels/LoggingChannelFactory.java]: Changed.

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java (hunk 3 above)
Comment (jdcove2) [reply src/main/java/emissary/core/channels/LoggingChannelFactory.java]: Changed.

File: src/main/java/emissary/core/channels/LoggingChannelFactory.java (hunk 5 above)
Comment (jdcove2) [reply src/main/java/emissary/core/channels/LoggingChannelFactory.java]: Changed to 'o' to be similar to the other comments.

**Review by drivenflywheel — APPROVED** [review]
LGTM

**Review by fbruton — APPROVED** [review]

**Pull request merged**